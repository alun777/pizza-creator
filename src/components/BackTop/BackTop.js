import React, { Component } from 'react';
import { connect } from 'react-redux';
import { actionCreators } from './store/index';

export class BackTop extends Component {

  render() {
    const { showScrollTop } = this.props
    return (
      <button
        className={showScrollTop ? 'backTop backTop__show' : 'backTop backTop__hide'}
        type="button"
        onClick={this.handleScrollTop}
      >
        <i className="backTop__iconfont">&#xe62e;</i>
      </button>
    )
  }

  componentDidMount() {
    this.bindEvents();
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.props.changeScrollTop)
  }

  bindEvents() {
    window.addEventListener('scroll', this.props.changeScrollTop)
  }

  handleScrollTop(event) {
    event.preventDefault();
    window.scrollTo(0, 0);
  }
}

export const mapStateToProps = (state) => {
  return {
    showScrollTop: state.getIn(['BackTop', 'showScrollTop'])

  }
}

export const mapDispatchToProps = (dispatch) => {
  return {
    changeScrollTop() {
      if (document.documentElement.scrollTop > 200) {
        dispatch(actionCreators.toggleScrollTop(true))
      } else {
        dispatch(actionCreators.toggleScrollTop(false))
      }
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(BackTop);
