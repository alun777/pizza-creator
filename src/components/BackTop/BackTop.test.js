import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import { BackTop, mapStateToProps, mapDispatchToProps } from './BackTop';

const props = {
  showScrollTop: true
}

describe('testing mapStateToProps and mapDispatchToProps', () => {
  it('should show the correct showScrollTop value', () => {
    const initialState = fromJS({
      BackTop: {
        showScrollTop: true
      }
    })
    expect(mapStateToProps(initialState).showScrollTop).toEqual(true)
  })

  it('should dispatch the correct action', () => { 
    const dispatch = jest.fn();
    
    mapDispatchToProps(dispatch).changeScrollTop();

    expect(dispatch).toHaveBeenCalled();
  })
})


let wrapper
beforeEach(() => {
  wrapper = shallow(<BackTop {...props} />)
})

it('should have a correct iconfont element', () => {
  expect(wrapper.find('button').length).toBe(1);

  expect(wrapper.find('button').text()).toBeTruthy();
})

it('successfully calls the onClick handler', () => {
  expect(wrapper.find('button').length).toEqual(1)
  window.scrollTo = jest.fn()
  let prevented = false
  wrapper.find('button').simulate('click', {
    preventDefault: () => {
      prevented = true;
    },
  });
  expect(prevented).toBe(true);
  wrapper.find('button').simulate('click', {
    preventDefault: jest.fn()
  })
  expect(window.scrollTo).toHaveBeenCalled()
})

it('should match snapshot', () => {
  expect(wrapper).toMatchSnapshot()
})
