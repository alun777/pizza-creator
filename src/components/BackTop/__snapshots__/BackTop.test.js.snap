// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<button
  className="backTop backTop__show"
  onClick={[Function]}
  type="button"
>
  <i
    className="backTop__iconfont"
  >
    î˜®
  </i>
</button>
`;
